### PR TITLE
Kitty Items Documentation update

### PR DESCRIPTION
**Pull request by MrDSGC**
- Updated documentation for api directory
- Updated documentation for web directory
- Added documentation of .ki-scripts per suggestion #239

### CODE REVIEW

File: api/src/workers/base-event-handler.ts [hunk 1]
```diff
@@ -41,6 +42,13 @@ abstract class BaseEventHandler {
       if (fromBlock > toBlock) {
         fromBlock = toBlock;
       }
+
+      // getEventsAtBlockHeightRange() has a block limit of 250 blocks. 
+      // If the range exceeds the limit of 200 then hard cap the block search range
+      if (fromBlock + this.stepSize < toBlock) {
+        toBlock = fromBlock + this.stepSize;
+      }
+
```
Comment (psiemens) [inline api/src/workers/base-event-handler.ts:51]: Do you mind moving this change into a separate PR from the documentation updates? Best to keep it isolated.

File: api/package.json [hunk 2]
```diff
@@ -44,6 +44,6 @@
     "ts-node-dev": "^1.1.1"
   },
   "engines": {
-    "node": "16.x"
+    "node": ">=16.x"
```
Comment (psiemens) [inline api/package.json:47]: Same here, I'd bump this in a separate PR.

**Review by psiemens — COMMENTED** [review]
Thanks for updating the documentation! I'll need a bit more time to read through it all. In the meantime left some comments on the code changes.

File: README.md [hunk 3]
```diff
@@ -42,8 +45,8 @@ From the `./web` directory run `npm install -D @next/swc-darwin-arm64`
 1.  Run `npm run dev:emulator`
 
     - Local development uses the [Flow Emulator](https://docs.onflow.org/emulator/) and the [FCL Development Wallet](https://github.com/onflow/fcl-dev-wallet) to simulate the blockchain and an FCL-compatible wallet.
+    - Running this npm script initializes your application by completing multiple important processes found [here](/.ki-scripts/README.md)
 
-2.  In another terminal, from the root of the project run `flow dev-wallet`
 
 Thats it! 🏁
```
Comment (psiemens) [inline README.md:51]: It probably doesn't make sense to use a numbered list now that there's only 1 step.

I'd recommend starting with a simple instruction and then link out to the scripts README afterwards. This minimizes the amount of text the developer needs to read.

I also find it better to put commands in full code blocks (i.e. three back ticks) because it's easier to copy.

---

## Local development

Run this command to start Kitty Items with the Flow local development suite:

```sh
npm run dev:emulator
```

Note: this script launches the Flow Emulator and FCL Development Wallet using the [initialization code in this repository](/.ki-scripts/README.md).

File: api/README.md [hunk 4]
```diff
@@ -1,52 +1,73 @@
 # Kitty Items API
+The Kitty Items API is a RESTful API built with [express](https://expressjs.com/) that sends transactions to Flow using the [Flow JS SDK](https://github.com/onflow/fcl-js/tree/master/packages/sdk). It contains endpoints for both the [kitty-items](src/services/kitty-items.ts) & [storefront](src/services/storefront.ts) services to read & write data from both the Flow blockchain as well as a SQL DB.
 
-The Kitty Items API is a RESTful API built with [express](https://expressjs.com/) that sends transactions to Flow using the [Flow JS SDK](https://github.com/onflow/fcl-js/tree/master/packages/sdk).
+## Services
+Kitty Items uses main 4 services to function. These services are used to maintain data integrity between the blockchain & our database as well as allow avenues for users to interact with our data through exposed api endpoints.
```
Comment (psiemens) [inline api/README.md:5]:
```suggestion
The Kitty Items API contains four main services. These services synchronize data between the blockchain and the SQL database, and also provide API endpoints consumed by the Kitty Items client-side application. 
```

File: api/README.md [hunk 5]
```diff
@@ -1,52 +1,73 @@
...
+The Kitty Items API is a RESTful API built with [express](https://expressjs.com/) that sends transactions to Flow using the [Flow JS SDK](https://github.com/onflow/fcl-js/tree/master/packages/sdk). It contains endpoints for both the [kitty-items](src/services/kitty-items.ts) & [storefront](src/services/storefront.ts) services to read & write data from both the Flow blockchain as well as a SQL DB.
 
-The Kitty Items API is a RESTful API built with [express](https://expressjs.com/) that sends transactions to Flow using the [Flow JS SDK](https://github.com/onflow/fcl-js/tree/master/packages/sdk).
+## Services
+Kitty Items uses main 4 services to function. These services are used to maintain data integrity between the blockchain & our database as well as allow avenues for users to interact with our data through exposed api endpoints. 
 
-## Setup
+### [Flow Service](src/services/flow.ts)
+This service contains functions that interact with the Flow blockchain using the [onflow/fcl](https://docs.onflow.org/fcl/) package. While it has no exposed endpoints, the methods here used to read, write & authorize data on the chain are used extensively in its sister services & workers.
```
Comment (psiemens) [inline api/README.md:8]:
```suggestion
This service contains functions that interact with the Flow blockchain using the [FCL JS](https://docs.onflow.org/fcl/) library. While it has no exposed endpoints, its methods are used to read, write and authorize data on the chain are used extensively by its sister services and workers.
```

IMO best to only use ampersands (&) in titles

File: web/README.md [hunk 6]
```diff
@@ -1,34 +1,20 @@
-This is a [Next.js](https://nextjs.org/) project bootstrapped with [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app).
+# Kitty Items Web
 
-## Getting Started
+Kitty Items Web is a [Next.js](https://nextjs.org/) project bootstrapped to serve 4 main views to allow users to interface with the Flow blockhain. Users can mint Kitty Items through an admin view, purchase through an item page, browse the marketplace, & view their Kitty Items on their profile. Users can navigate between these pages through a navigation bar at the top of their screen.
```
Comment (psiemens) [inline web/README.md:3]:
```suggestion
The Kitty Items web app is a [Next.js](https://nextjs.org/) project bootstrapped to serve four main views and allow users to interface with the Flow blockchain. Users can mint Kitty Items through an admin view, purchase through an item page, browse the marketplace, and view their Kitty Items on their profile. Users can navigate between these pages through a navigation bar at the top of their screen.
```

File: web/README.md [hunk 7]
```diff
@@ -1,34 +1,20 @@
...
 
-First, run the development server:
+## [Homepage View](pages/index.jsx)
+![Homepage](/assets/kitty-items-homepage-view.png)
+The homepage shows the most recent listings for both the Storefront and Marketplace.
```
Comment (psiemens) [inline web/README.md:7]:
```suggestion
The homepage shows the most recent listings for both the storefront and marketplace.
```

I don't think "storefront" and "marketplace" are proper nouns in this case

File: web/README.md [hunk 8]
```diff
@@ -1,34 +1,20 @@
...
-
-## Deploy on Vercel
-
-The easiest way to deploy your Next.js app is to use the [Vercel Platform](https://vercel.com/new?utm_medium=default-template&filter=next.js&utm_source=create-next-app&utm_campaign=create-next-app-readme) from the creators of Next.js.
-
-Check out our [Next.js deployment documentation](https://nextjs.org/docs/deployment) for more details.
+## [Admin minter view](pages/admin/mint.jsx)
+![Admin](/assets/kitty-items-admin-view.png)
+The admin view is where users can mint Kitty Items through the admin service account. Kitty Items minted here are automatically listed for purchase by the admin account. Unpurchased 
+Kitty Items will go to the Storefront.
```
Comment (psiemens) [inline web/README.md:20]:
```suggestion
The admin view is where users can mint new NFTs through the Kitty Items minter account. It's open to anybody for demonstration purposes. Kitty Items minted here are automatically put up for sale on the storefront page.
```

File: api/README.md [hunk 9]
```diff
@@ -1,52 +1,73 @@
...
-  --url http://localhost:3000/v1/kitty-items/setup \
-  --header 'Content-Type: application/json'
-```
+Notable functions:
+- `findOrCreateLatestBlockCursor()`: Queries the `block_cursor` table for the the last block check by the listing handler worker. If no row is found then insert a record with our current latest block height.
+- `updateBlockCursorById()`: Updates the row with the latest block height that has been checked for events.
 
-- **POST /v1/market/setup** - Create a resource that allows the `MINTER_FLOW_ADDRESS` to hold listings for Kitty Items.
+### [Kitty-items Service](src/services/kitty-items.ts)
+This service contains functions that utilize the Flow Service to send/execute transactions/scripts that fetch/manipulate Kitty Items data. These functions are exposed through endpoints to allow users to read & manipulate Kitty Items data on the chain. The Cadence scripts/transactions can also be called in the [Flow-cli](https://docs.onflow.org/flow-cli/) and directly read/write to the Flow blockchain.
```
Comment (psiemens) [inline api/README.md:24]:
```suggestion
This service contains functions that utilize the Flow Service to send Cadence transactions and scripts that fetch and update Kitty Items data.

You can also run the [transactions and scripts](../cadence) manually using the [Flow CLI](https://docs.onflow.org/flow-cli/).
```

I'd keep this simple.

File: api/README.md [hunk 10]
```diff
@@ -1,52 +1,73 @@
...
 
-## Minting
+- **POST `/v1/kitty-items/mint`** *[DEPRECATED]*: Mint a Kitty Item and send it to a recipient account. Calls [mint_kitty_item.cdc](/cadence/transactions/kittyItems/mint_kitty_item.cdc)
 
-Run the commands below to mint new items and list them for sale.
+- **POST `/v1/kitty-items/mint-and-list`**: Mint a Kitty Item and send it to a recipient account. Create a listing to sell the Kitty Item immediately after. Called by the admin service account to mint & list items onto the store. Calls [mint_and_list_kitty_item.cdc](/cadence/transactions/kittyItems/mint_and_list_kitty_item.cdc)
 
-- **POST /v1/kitty-items/mint** - Mint a Kitty Item
-  and send it to the `recipient` account.
+- **POST `/v1/kitty-items/transfer`**: Sends a Kitty Item from account to recipient account. Calls [transfer_kitty_item.cdc](/cadence/transactions/kittyItems/transfer_kitty_item.cdc)
```
Comment (psiemens) [inline api/README.md:33]:
```suggestion
- **POST `/v1/kitty-items/transfer`**: Sends a Kitty Item from the admin account to a recipient account. Calls [transfer_kitty_item.cdc](/cadence/transactions/kittyItems/transfer_kitty_item.cdc)
```

File: api/README.md [hunk 11]
```diff
@@ -1,52 +1,73 @@
...
-  --header 'Content-Type: application/json'
-```
+[Exposed Endpoints](src/routes/kitty-items.ts):
+- **POST `/v1/kitty-items/setup`** *[DEPRECATED]*: Sets up account with a resource that stores Kitty Items in the admin minter account. The [startup script](/.ki-scripts/startup.js) directly sends the setup Cadence transaction instead of using this endpoint. Calls [setup_account.cdc](/cadence/transactions/kittyItems/setup_account.cdc)
 
-## Minting
+- **POST `/v1/kitty-items/mint`** *[DEPRECATED]*: Mint a Kitty Item and send it to a recipient account. Calls [mint_kitty_item.cdc](/cadence/transactions/kittyItems/mint_kitty_item.cdc)
 
-Run the commands below to mint new items and list them for sale.
+- **POST `/v1/kitty-items/mint-and-list`**: Mint a Kitty Item and send it to a recipient account. Create a listing to sell the Kitty Item immediately after. Called by the admin service account to mint & list items onto the store. Calls [mint_and_list_kitty_item.cdc](/cadence/transactions/kittyItems/mint_and_list_kitty_item.cdc)
```
Comment (psiemens) [inline api/README.md:31]:
```suggestion
- **POST `/v1/kitty-items/mint-and-list`**: Mint a Kitty Item and immediately list it for sale. Called by the admin account to mint and sell new items. Calls [mint_and_list_kitty_item.cdc](/cadence/transactions/kittyItems/mint_and_list_kitty_item.cdc)
```

I would just say "admin account" instead of "admin service account" because "service account" doesn't mean much outside the context of the emulator.

File: api/README.md [hunk 12]
```diff
@@ -1,52 +1,73 @@
...
-curl --request POST \
-  --url http://localhost:3000/v1/kitty-items/setup \
-  --header 'Content-Type: application/json'
-```
+Notable functions:
+- `findOrCreateLatestBlockCursor()`: Queries the `block_cursor` table for the the last block check by the listing handler worker. If no row is found then insert a record with our current latest block height.
+- `updateBlockCursorById()`: Updates the row with the latest block height that has been checked for events.
 
-- **POST /v1/market/setup** - Create a resource that allows the `MINTER_FLOW_ADDRESS` to hold listings for Kitty Items.
+### [Kitty-items Service](src/services/kitty-items.ts)
```
Comment (psiemens) [inline api/README.md:23]:
```suggestion
### [Kitty Items Service](src/services/kitty-items.ts)
```

File: api/README.md [hunk 13]
```diff
@@ -1,52 +1,73 @@
...
-Run the commands below to initialize the minter account.
+Notable functions:
+- `sendTx()`: Takes a Cadence transaction as an arguement and sends it to write data into the Flow blockchain.
+- `executeScript()`: Takes a Cadence script as an arguement and executes it to read data from the Flow blockchain. 
+- `getLatestBlockHeight()`: Gets the most recent block from the Flow blockchain.
+- `authorizeMinter()`: Returns an asynchronous method to authorize an account to mint Kitty Items.
 
-- **POST /v1/kitty-items/setup** - Create a resource that holds Kitty Items in the `MINTER_FLOW_ADDRESS` account.
+### [Block-cursor Service](src/services/block-cursor.ts)
+This service contains functions that read & write data to the SQL database to maintain the Kitty Items Marketplace
```
Comment (psiemens) [inline api/README.md:17]:
```suggestion
This service watches for state-changing events emitted by the Kitty Items contract and updates the SQL database accordingly.
```

File: api/README.md [hunk 14]
```diff
@@ -1,52 +1,73 @@
...
 
-Run the commands below to initialize the minter account.
+Notable functions:
+- `sendTx()`: Takes a Cadence transaction as an arguement and sends it to write data into the Flow blockchain.
+- `executeScript()`: Takes a Cadence script as an arguement and executes it to read data from the Flow blockchain. 
+- `getLatestBlockHeight()`: Gets the most recent block from the Flow blockchain.
+- `authorizeMinter()`: Returns an asynchronous method to authorize an account to mint Kitty Items.
 
-- **POST /v1/kitty-items/setup** - Create a resource that holds Kitty Items in the `MINTER_FLOW_ADDRESS` account.
+### [Block-cursor Service](src/services/block-cursor.ts)
```
Comment (psiemens) [inline api/README.md:16]:
```suggestion
### [Block Cursor Service](src/services/block-cursor.ts)
```

File: api/README.md (hunk 12 above)
Comment (psiemens) [reply api/README.md:23]: Also, I noticed that you took out the sample curl commands. I think those were helpful to show developers how to call the API.

File: .ki-scripts/README.md [hunk 15]
```diff
@@ -0,0 +1,45 @@
+# Kitty Items Scripts
+
```
Comment (10thfloor) [inline .ki-scripts/README.md:2]: Can we add a short description of what this document is all about? It's not obvious when jumping into it without context.

File: .ki-scripts/README.md [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+## Version Check
+
+A script that compares the current process's Node version against the [package.json](/package.json) Node engine requirement.
+
+## Startup
+
+This script is called when deploying both local & testnet versions of the Kitty Items application. Multiple processes are executed to initialize both the database as well as an admin service account required for the application to function.
+
+### Begin Process
+The script begins by ending all running processes from previously launched applications.
```
Comment (10thfloor) [inline .ki-scripts/README.md:12]:
```suggestion
The script begins by ending all running processes from previous runs of the script.
```

File: .ki-scripts/README.md [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+
+### Check for correct Node version
+In this step, the script checks the process's Node version against the Node engine requirement in the root [package.json](/package.json). Similar to [version-check.js](/.ki-scripts/version-check.js).
```
Comment (10thfloor) [inline .ki-scripts/README.md:15]: Why is this done twice?

File: .ki-scripts/README.md [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+### Testnet account creation (`bootstrapNewTestnetAccount()`)
+This section defines the function that begins the process of creating a new Testnet account for applications running on the testnet environment. This account becomes the admin service account that eventually mints your Kitty Items on the testnet.
```
Comment (10thfloor) [inline .ki-scripts/README.md:18]:
```suggestion
This section defines the function that begins the process of creating a new Testnet account for applications that connect to and run Cadence on the testnet. This account becomes the admin service account that eventually mints your Kitty Items on the testnet.
```

Comment (10thfloor) [reply .ki-scripts/README.md:18]: I'm still not sure whether to use "the testnet" or Testnet ... but I think we should pick one per doc and stick to it, instead of having both. This is a minor complaint, but it bugs me.

File: .ki-scripts/README.md [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+
+### Contract deployment (`deployAndInitialize()`)
+This section defines the function that begins the process of deploying the required cadence contracts to the admin service account. The contracts being deployed can be found in [flow.json](/flow.json).
+
+- **Initialize Kitty Items**: This step executes the [setup_account.cdc](/cadence/transactions/kittyItems/setup_account.cdc) transaction and initializes the admin service account with the resources required to mint & transfer Kitty Items.
+
+- **Initialize NFTStorefront**: This step executes the [setup_account.cdc](/cadence/transactions/nftStorefront/setup_account.cdc) transaction and initializes the admin service account with the resources required to create Kitty Item listings on the marketplace.
+
+### Emulator environment startup
+This process initializes the Flow emulator and creates a local instance of the Flow blockchain to develop against. The emulator will also initialize an instance of the FCL Dev Wallet. Once the processes are running, the script will execute the `Contract deployment` steps to set up the admin service account.
```
Comment (10thfloor) [inline .ki-scripts/README.md:28]:
```suggestion
This process initializes the Flow emulator which starts a local instance of the Flow blockchain protocol to develop against. The emulator will also initialize an instance of the [FCL Dev Wallet](https://github.com/onflow/fcl-dev-wallet). Once the processes are running, the script will execute the `Contract deployment` steps to set up the admin service account.
```

**Review by 10thfloor — CHANGES_REQUESTED** [review]

File: api/README.md (hunk 5 above)
Comment (MrDSGC) [reply api/README.md:8]: `While it has no exposed endpoints, its methods ~are~ used to read, write and authorize data on the chain are used extensively by its sister services and workers.`

File: .ki-scripts/README.md (hunk 17 above)
Comment (MrDSGC) [reply .ki-scripts/README.md:15]: the stand-alone version check script is ran through `npm run preinstall` 
the logic is also in the startup script. adding a line to the `Version Check` description on how to run the preinstall script

File: .ki-scripts/README.md (hunk 18 above)
Comment (MrDSGC) [reply .ki-scripts/README.md:18]: going with Testnet

File: api/README.md (hunk 12 above)
Comment (MrDSGC) [reply api/README.md]: i took out the curl commands since we had flow scripts that sort of abstract away the endpoints but i can see the value in letting them hit the endpoints themselves. ill add the curl commands back in a `Getting Started` section

**Review by 10thfloor — APPROVED** [review]
